src/array_texture_preload.rs

```diff
@@ -0,0 +1,70 @@
+use std::sync::{Arc, RwLock};
+
+use bevy::{
+    prelude::{Assets, Handle, Image, Res, ResMut},
+    render::{render_resource::FilterMode, texture::ImageSettings, Extract},
+};
+
+use crate::{
+    prelude::{TilemapSpacing, TilemapTileSize},
+    render::TextureArrayCache,
+};
+
+#[derive(Default, Debug, Clone)]
+pub struct TilemapArrayTexture {
+    pub atlas_texture: Handle<Image>,
+    pub tile_size: TilemapTileSize,
+    pub tile_spacing: TilemapSpacing,
+    /// Defaults to ImageSettings.
+    pub filter: Option<FilterMode>,
+}
+
+/// A bevy world resource that allows you to add atlas textures for
+/// loading into the array texture cache.
+#[derive(Default, Debug, Clone)]
+pub struct ArrayTextureLoader {
+    // Arc and RwLock just let us get around Extract's read only rules.
+    textures: Arc<RwLock<Vec<TilemapArrayTexture>>>,
+}
+
+impl ArrayTextureLoader {
+    pub fn add(&self, texture: TilemapArrayTexture) {
+        if let Ok(mut textures) = self.textures.try_write() {
+            textures.push(texture);
+        }
+    }
+
+    pub(crate) fn drain(&self) -> Vec<TilemapArrayTexture> {
+        if let Ok(mut textures) = self.textures.try_write() {
+            return textures.drain(..).collect::<Vec<_>>();
```

Comment:
uhm ... why? why copy all the stuff from one Vec into a new Vec to return? This can be much more efficiently done by taking the Vec out and leaving a default empty Vec in its place:
```suggestion
            return std::mem::take(&mut *textures);
```

Reply:
I typically shy away from `std::mem` but in this case I think it makes sense. 👍